### DISCUSSION TITLE
How to Model this behavior

### DISCUSSION THREAD

**Discussion by seloner** [statelyai/xstate#2244] [Q&A]
Hello fans of xstate ! 
Lets say we have a form machine and a set of plugins
User needs to be able to change the form and the plugins extend the behavior of the form.
For example we have 2 plugins(file plugin,and printPdf plugin) . one is doing file upload after the submit of the form and the other is the doing the print pdf job . I am thinking plugins as actors .

Important:We have a config file that tells us which plugins to use so maybe we have 1 plugin to run maybe 2 or maybe 0 .

The big picture is like this 
![image](https://user-images.githubusercontent.com/29483115/119776786-0a488f80-bece-11eb-8169-00befb04c4a1.png)

Plugins need to finish in order for the form to be done .

So one solution is to spawn an actor to execute the plugins. This actor will spawn all plugins actors and also needs to wait them to finish . 

this would look like this 
![image](https://user-images.githubusercontent.com/29483115/119778056-ad4dd900-becf-11eb-931a-d50ffb665448.png)

The main issue how to track that all plugins finished their job .

![image](https://user-images.githubusercontent.com/29483115/119778481-441a9580-bed0-11eb-9bd4-09ac94f7066f.png)

Maybe have a registered plugins array in context and track if they are finished based on an always transition ?
Maybe executing plugins machine should be a parallel machine ?
What do you think ?

**Comment by VanTanev** [comment]
I've done the invoke-actors solution in the past. It gets quite complex quite fast, so I would recommend trying to avoid it if possible. Here is my suggestion:

1) If the list of plugins is not static, they should be set on the context as `{ plugins: Array<Plugin> }`
2) I'm assuming plugins are async, and can fail
3) I'm assuming plugins can execute in parallel

a "simple" solution would be something like:
```typescript
const machine = createMachine({
  initial: 'idle',
  states: {
    idle: {
      on: { 
        SUBMIT: { target: 'executing' }
      }
    },
    submitting: {
      invoke: {
        src: () => submit(),
        onDone: { target: 'executingPlugins' }
      }
    },
    executingPlugins: {
      invoke: {
        src: (ctx) => Promise.all(ctx.plugins.map(plugin => plugin.execute(ctx))
        onDone: { target: 'finished' },
        onError: {
           target: 'idle',
           actions: (_, e) => showErrorNotification(e.data)
        }
      }
    },
    finished: { type: final }
  }
})
```

The exact form of the `executingPlugins` invoke can be different, but I would suggest not doing spawned actors for them, but doing a single invoke that can return either success or failure.

Spawned actors are necessary when you want to communicate with said actors; If it's just a bunch of tasks that can either succeed or fail, there is little need for spawning; a normal promise chain is simpler.

Also, depending on how you want this to behave, you might want to use `Promise.allSettled()`, and track which plugins succeeded and which failed, and offer a RETRY event to retry only the ones that failed.

**Reply by seloner** [reply]
@VanTanev Thanks for your answer.I like your solution but I forgot to mention that some plugins need to hold state also.
For example I need a transaction plugin . One field of the form is a transaction code . User enters the code in the input of the form and I need call some api services to gather information about the code and show a checkmark if the code is valid .
I guess I could hold the state for the code in the form machine and just use the plugin to call the apis and check if the code is valid or not but then I feel the form machine holds in the context a lot .
Thats why I wanted to use actors in order to split the state and the complexity . But the issue that is hard to distribute and share states between actors .At least thats what I have experienced so far .

**Reply by VanTanev** [reply]
Yeah, seems like you're in "no easy solutions" land.

If you were using spawned machines, you could take some ideas from this (I have not modified the example from the original code, but hopefully you can follow it)

```typescript
import { createMachine, assign } from 'xstate'
import { WIPMachineActor } from 'shared/resource/WIP.machine'
import { WIPDataType } from 'model/WIP'

export type SyncWIPsGuard = (ctx: SyncWIPsContext, e: any) => boolean

export interface SyncWIPsContext {
    wipRefs: {
        status: 'pending' | 'aquired' | 'failed' | 'released'
        ref: WIPMachineActor<any>
    }[]
}
export type SyncWIPsEvent = {
    type: 'WIP_SYNC_STATUS'
    data: {
        dataType: WIPDataType
        status: 'aquired' | 'failed' | 'released'
    }
}
export const syncWIPsMachine = createMachine<SyncWIPsContext, SyncWIPsEvent>(
    {
        strict: true,
        on: {
            WIP_SYNC_STATUS: { actions: 'syncWipStatus' },
        },
        initial: 'precondition',
        states: {
            precondition: {
                always: [
                    {
                        cond: 'allInStartCondition',
                        target: 'executing',
                    },
                    { target: 'failed' },
                ],
            },
            executing: {
                entry: 'sendSyncedSignal',
                always: [
                    {
                        cond: 'failedMachineExists',
                        target: 'cleanupAfterFailure',
                    },

                    {
                        cond: 'allInTargetCondition',
                        target: 'done',
                    },
                ],
            },
            cleanupAfterFailure: {
                entry: 'handleFailedMachines',
                always: {
                    cond: 'cleanupComplete',
                    target: 'failed',
                },
            },
            failed: { type: 'final', data: () => false },
            done: { type: 'final', data: () => true },
        },
    },
    {
        guards: {
            failedMachineExists: ctx =>
                // all machines executed
                ctx.wipRefs.every(wipRef => wipRef.status !== 'pending') &&
                // and at least one failed
                ctx.wipRefs.some(wipRef => wipRef.status === 'failed'),
        },
        actions: {
            syncWipStatus: assign((ctx, e) => ({
                wipRefs: ctx.wipRefs.map(wipRef => {
                    if (wipRef.ref.state.context.dataType === e.data.dataType) {
                        wipRef.status = e.data.status
                    }
                    return wipRef
                }),
            })),
        },
    },
)
```

```typescript
// multi-lock aquire machine
const allMachinesInReading: SyncWIPsGuard = ctx =>
    ctx.wipRefs.every(wipRef => wipRef.ref.state.matches('reading'))
const allMachinesAquired: SyncWIPsGuard = ctx =>
    ctx.wipRefs.every(wipRef => wipRef.status === 'aquired')
const multiLockAquireMachine = syncWIPsMachine.withConfig({
    guards: {
        cleanupComplete: ctx =>
            ctx.wipRefs.every(wipRef => ['failed', 'released'].includes(wipRef.status)),
        allInStartCondition: allMachinesInReading,
        allInTargetCondition: allMachinesAquired,
    },
    actions: {
        sendSyncedSignal: actions.pure(ctx =>
            ctx.wipRefs.map(wipRef => send('ATTEMPT_AQUIRE', { to: () => wipRef.ref })),
        ),
        handleFailedMachines: actions.pure(ctx =>
            ctx.wipRefs
                .filter(wipRef => wipRef.status === 'aquired')
                .map(wipRef => send('RELEASE', { to: () => wipRef.ref })),
        ),
    },
})
```

```typescript
 // in main machine
                aquiringLocks: {
                    on: { WIP_SYNC_STATUS: { actions: forwardTo('sync-wips-machine') } },
                    invoke: {
                        id: 'sync-wips-machine',
                        src: multiLockAquireMachine,
                        data: (ctx, e): SyncWIPsContext => {
                            invariant(e.type === 'START_EDITING')
                            const wipRefs: WIPMachineActor<any>[] = e.selection.map(dataType => {
                                const ref = ctx.wips.find(
                                    wip => wip.state.context.dataType === dataType,
                                )
                                invariant(ref, `No WIP ref matching ${dataType}`)
                                return ref
                            })
                            return {
                                wipRefs: wipRefs.map(ref => ({ status: 'pending', ref })),
                            }
                        },
                        onDone: [
                            {
                                // if multi-lock finished with "true" for success
                                cond: (_, e) => !!e.data,
                                target: 'editing',
                            },
                            {
                                target: 'idle',
                                actions: 'notificationCouldNotAquireLocks',
                            },
                        ],
                    },
                },
```

**Reply by seloner** [reply]
Nice idea . I think another solution its to have form state outside xstate in a store like zustand  and you use machines only to control the ui and complex api calls and use stores for read writes .
